src/main/java/com/metamx/emitter/core/HttpPostEmitter.java

```diff
@@ -166,315 +190,344 @@ public void emit(Event event)
+    }
+    catch (TimeoutException e) {
+      String message = String.format("Timed out after [%d] millis during flushing", config.getFlushTimeOut());
+      throw new IOException(message, e);
+    }
+    catch (InterruptedException e) {
+      log.debug("Thread Interrupted");
+      Thread.currentThread().interrupt();
+      throw new IOException("Thread Interrupted while flushing", e);
     }
   }
 
   @Override
   @LifecycleStop
   public void close() throws IOException
   {
-    synchronized (started) {
-      // flush() doesn't do things if it is not started, so flush must happen before we mark it as not started.
-      flush();
-      started.set(false);
-      exec.shutdown();
+    synchronized (startLock) {
+      if (started) {
+        started = false;
+        Batch lastBatch = concurrentBatch.getAndSet(null);
+        flush(lastBatch);
+        // Emitter is interrupted after the last batch is flushed.
+        emittingThread.interrupt();
+      }
     }
   }
 
-  private class EmittingRunnable implements Runnable
+  private class EmittingThread extends Thread
   {
-    private final long instantiatedVersion;
+    private ZeroCopyByteArrayOutputStream gzipBaos;
 
-    public EmittingRunnable(
-        long instantiatedVersion
-    )
+    EmittingThread()
     {
-      this.instantiatedVersion = instantiatedVersion;
+      super("HttpPostEmitter-" + instanceCounter.incrementAndGet());
+      setDaemon(true);
     }
 
     @Override
     public void run()
     {
-      long currVersion = version.get();
-
-      try {
-        if (!started.get()) {
-          log.info("Not started, skipping...");
-          return;
+      while (true) {
+        boolean needsToShutdown = needsToShutdown();
+        try {
+          emitLargeEvents();
+          emitBatches();
         }
-
-        if (instantiatedVersion != currVersion) {
-          log.debug("Skipping because instantiatedVersion[%s] != currVersion[%s]", instantiatedVersion, currVersion);
+        catch (Throwable t) {
+          log.error(t, "Uncaught exception in EmittingThread.run()");
+        }
+        if (needsToShutdown) {
+          // Make GC life easier
+          drainBuffersToReuse();
           return;
-        } else {
-          count.set(0);
-          currVersion = version.incrementAndGet();
         }
+        // Waiting for 1/2 of config.getFlushMillis() in order to flush events not more than 50% later than specified.
+        // If nanos=0 parkNanos() doesn't wait at all, that we don't want.
+        long waitNanos = Math.max(TimeUnit.MILLISECONDS.toNanos(config.getFlushMillis()) / 2, 1);
+        LockSupport.parkNanos(HttpPostEmitter.this, waitNanos);
+      }
+    }
 
-        final List<byte[]> events;
-        synchronized (eventsList) {
-          events = eventsList.getAndSet(Lists.<byte[]>newLinkedList());
+    private boolean needsToShutdown()
+    {
+      boolean needsToShutdown = Thread.interrupted();
```

Comment:
interrupted flag will be erased if InterruptedException is thrown and caught for instance within sendWithRetry.
It might not affect the logic though. But then it is not clear why do we need `emittingThread.interrupt()` and check it here.